pkg/dockerregistry/server/auth.go

```diff
@@ -272,6 +273,59 @@ func verifyOpenShiftUser(ctx context.Context, client client.UsersInterface) erro
 	return nil
 }
 
+// authorizeRepositoryAccess returns nil if the given client is granted access to the given
+// <namespace>/<imageRepo> repository. Otherwise the access is denied.
+func authorizeRepositoryAccess(ctx context.Context, namespace, imageRepo, verb string, client client.LocalSubjectAccessReviewsNamespacer) error {
+	if isCrossRepoMountFromSourceRepository(ctx, namespace, imageRepo, verb) {
+		return nil
+	}
+
+	return verifyImageStreamAccess(ctx, namespace, imageRepo, verb, client)
+}
+
+// isCrossRepoMountFromSourceRepository handles a case of accessing source repository requested by a cross
+// repository mount during a blob upload. During a push with the cross-repo mount, there are two authorization
+// checks:
+//
+//   1. repository:<source-namespace>/<source-name>:pull
+//   2. repository:<target-namespace>/<target-name>:push
+//
+// This function always grants access in the first case because OpenShift registry treats all the local blobs
+// as accessible once authorized to access given repository (target one).
+func isCrossRepoMountFromSourceRepository(ctx context.Context, namespace, imageRepo, verb string) bool {
+	if verb != "get" {
+		return false
+	}
+	if method := strings.ToLower(getStringValueFromContext(ctx, "http.request.method", "")); method != "post" {
+		return false
+	}
+
+	uriString := getStringValue(ctx, "http.request.uri")
+	uri, err := url.Parse(uriString)
+	if err != nil {
+		context.GetLogger(ctx).Warnf("failed to parse requested uri %q: %v", uri, err)
+		return false
+	}
+	from, exists := uri.Query()["from"]
```

Comment:
> 1. where is the access check done prior to this function being called?

It's not done prior, it's done afterwards with a check for target repository.

> 2. can the from value (or the uri it comes from) be spoofed, allowing the verifyImageStreamAccess check to be skipped?

 The code relies on the fact that `<verb>` doesn't match `METHOD` (method says `POST` which corresponds to `update`) and the repository matches the `from` argument. Both are determined by upstream code from request's method and uri. I don't see a way for a client to make up a request in a way to pass this function without being being a push with cross-repo mount (and therefor triggering another auth check for the target repository).